### ISSUE TITLE
Add support to SIARD 2 to the  Format Identifier in place (Siegfried)

### ISSUE THREAD

**Issue by jmaferreira** [keeps/roda#809]
https://tika.apache.org/1.1/detection.html

**Comment by hsilva-keep** [comment]
@jmaferreira format identification currently is done by siegfried and not Tika. In that sense, what is the purpose of this issue?

**Comment by jmaferreira** [comment]
Fixed.

**Comment by jmaferreira** [comment]
We must have a way to execute the action that ingests a series of databases from SIARD into the viewer with a single action. If the format is not well identified, this task is impossible. DGLAB requirement.

**Comment by hsilva-keep** [comment]
@jmaferreira this task clearly is related to siegfried & how it knows that a particular file is from a certain file format. And that is done via file signatures. In that sense, we need to search & decide on the best way to add to siegfried the ability to recognize SIARD (1 & 2) files.

Some tips/links: https://github.com/richardlehane/siegfried/wiki/Contributing-to-siegfried

Off-topic but related to your problem: you already can do what you want via advanced search & search by extension. from the result set, select them all & run plugin that sends information to the viewer.

**Comment by jmaferreira** [comment]
Yep.
>

**Comment by chalkos** [comment]
## Solution description

Solution: Extend the existing base droid signature file with a container file signature matcher that tries to find a file `header/metadata.xml` inside any ZIP container that contains the string `xmlns="http://www.bar.admin.ch/xmlns/siard/2.0/metadata.xsd"` between the first 4 bytes and 240 bytes of the file. The solution was extended to support SIARD 1.0, since it is a performance improvement over the existing SIARD 1.0 signature definition, and includes support for deflate compression method (which can be used in SIARD 1.0 but is not supported by the current SIARD 1.0 signature definition).

Files used in this solution:
* **siard_sig.xml** - standard signature file with the SIARD file format signature and PUID
* **siard_container.xml** - container signature file, specifying that the container is a ZIP archive that must contain a `header/metadata.xml` file, and that a specific string (the SIARD XML namespace) must be present in the beginning of that file.

## Solution details

This section contains details on how the solution is designed and how to install it.

### New files

The files in this section purposely use arbitrary IDs to avoid collisions with existing formats.

**siard_sig.xml**:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<FFSignatureFile DateCreated="2017-03-01T11:28:12+00:00" Version="1" xmlns="http://www.nationalarchives.gov.uk/pronom/SignatureFile">
  <FileFormatCollection>
    <FileFormat ID="3001" Name="SIARD (Software-Independent Archiving of Relational Databases)" PUID="dev/1" Version="1.0">
      <Extension>siard</Extension>
    </FileFormat>
    <FileFormat ID="3002" Name="SIARD (Software-Independent Archiving of Relational Databases)" PUID="dev/2" Version="2.0">
      <Extension>siard</Extension>
    </FileFormat>
  </FileFormatCollection>
</FFSignatureFile>
```

**siard_container.xml**:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ContainerSignatureMapping schemaVersion="1.0" signatureVersion="1">
  <ContainerSignatures>
    <ContainerSignature ContainerType="ZIP" Id="3001">
      <Description>SIARD 1.0 Metadata Namespace</Description>
      <Files>
        <File>
          <Path>header/metadata.xml</Path>
          <BinarySignatures>
            <InternalSignatureCollection>
              <InternalSignature ID="3001">
                <ByteSequence Reference="BOFoffset">
                  <SubSequence Position="1" SubSeqMaxOffset="256" SubSeqMinOffset="50">
                    <Sequence>'xmlns="http://www.bar.admin.ch/xmlns/siard/1.0/metadata.xsd"'</Sequence>
                  </SubSequence>
                </ByteSequence>
              </InternalSignature>
            </InternalSignatureCollection>
          </BinarySignatures>
        </File>
      </Files>
    </ContainerSignature>
    <ContainerSignature ContainerType="ZIP" Id="3002">
      <Description>SIARD 2.0 Metadata Namespace</Description>
      <Files>
        <File>
          <Path>header/metadata.xml</Path>
          <BinarySignatures>
            <InternalSignatureCollection>
              <InternalSignature ID="3002">
                <ByteSequence Reference="BOFoffset">
                  <SubSequence Position="1" SubSeqMaxOffset="256" SubSeqMinOffset="50">
                    <Sequence>'xmlns="http://www.bar.admin.ch/xmlns/siard/2.0/metadata.xsd"'</Sequence>
                  </SubSequence>
                </ByteSequence>
              </InternalSignature>
            </InternalSignatureCollection>
          </BinarySignatures>
        </File>
      </Files>
    </ContainerSignature>
  </ContainerSignatures>
  <FileFormatMappings>
    <FileFormatMapping Puid="dev/1" signatureId="3001"/>
    <FileFormatMapping Puid="dev/2" signatureId="3002"/>
  </FileFormatMappings>
  <TriggerPuids>
    <TriggerPuid ContainerType="ZIP" Puid="x-fmt/263"/>
  </TriggerPuids>
</ContainerSignatureMapping>
```

**NOTES:**
* The `Puid`s in both files must match (in this case, `dev/1` for SIARD 1.0 and `dev/2` for SIARD 2.0)
* In `siard_container.xml`, the attribute `Id` in element `ContainerSignature` must match the attribute `signatureId` in `FileFormatMapping` (in this case `3001` was used for SIARD 1.0 and `3002` for SIARD 2.0).
* In attribute `ID`, element `FileFormat`, `3001` and `3002` were used as IDs, but are unrelated to the `signatureId` (these numbers just need to be unique within a signature file).

### Building

You need siegfried and roy, instructions on how to obtain them are available here: https://github.com/richardlehane/siegfried/wiki/Getting-started#installing-siegfried

These instructions assume you are using linux (they may need some adjustments to work on MacOS).

1. Create a backup of `/usr/share/siegfried/default.sig` (just in case)

2. Save both the above XML files to a directory, which will be your siegfried home directory. This directory will be referred to as `SFHOME`.

3. (optional) Download and install the latest version of DROID and container signature files (instructions [here](https://github.com/richardlehane/siegfried/wiki/Building-a-signature-file-with-ROY#droid-and-container-signature-files)).

4. (optional) Run `roy harvest` to get the latest PRONOM reports.

5. Run `roy build -home "SFHOME" -extend "SFHOME/siard_sig.xml" -extendc "SFHOME/siard_container.xml" -name "siardsig" "SFHOME/default.sig"` to create a signature file based on the default signature files which also includes the new SIARD 1.0 and 2.0 file identification methods.

6. Run `sf -home "SFHOME" your_database.siard` to identify the file using the `default.sig` file in your siegfried home directory.

### Caveats:

- SIARD 2.0 files include an empty directory `header/version/2.0/` that was meant to simplify identifying SIARD files and their version, but siegfried can not recognise SIARD 2.0 files when the only rule given is

```xml
<File>
    <Path>header/version/2.0/</Path>
    <!-- using header/version/2.0 (without the last / ) does not work either -->
</File>
```

therefore the solution was to find the `metadata.xml` file and look for a specific XML namespace string.

## Improvements over the existing SIARD 1.0 format identifier

* The above rules correctly identify SIARD 1.0 files compressed using the DEFLATE method, which is supported, according to http://www.digitalpreservation.gov/formats/fdd/fdd000426.shtml :

> Although compression as supported by ZIP was originally not permitted, use of the "deflate" algorithm (the default compression option for ZIP) is supported in the SIARD software and was announced as a planned change in a [February 2015 presentation on SIARD](http://www.eark-project.com/resources/conference-presentations/dpc-eark-event/24-dpcsfasiard) from the Swiss Federal Archives.

* Performance improvements: Using a real-world SIARD 1.0 uncompressed 60GB database:

| Identification method | Time until identification as SIARD 1.0 | CPU usage |
|-----------------------|---------------------------|-----------|
| current               | 11 minutes                | 69%       |
| new (detailed above)  | 0.3 seconds               | 1%        |

## Further reading:

* http://openpreservation.org/blog/2016/01/07/droid-container-signature-files-what-they-are-and-how-to-create-them-a-template-and-an-example-or-few/
* https://github.com/richardlehane/siegfried/wiki

**Comment by dpreservation** [comment]
We'll be adding to PRONOM in our next release, due in March. Thank you.

**Comment by Dclipsham** [comment]
We'll be adding to PRONOM in our next release, due in March. Thank you.

**Comment by hsilva-keep** [comment]
This task is done and, as long as one uses RODA with Siegfried having a recent signature file (from March 2017 on), it will have SIARD 2 support in Siegfried (also available in RODA docker container version 2.0.0-beta5).

```bash
$> sf -version
siegfried 1.7.0
/usr/share/siegfried/default.sig (2017-03-12T11:55:35+11:00)
identifiers: 
  - pronom: DROID_SignatureFile_V89.xml; container-signature-20170308.xml
```
